[WIP] Add sync handler base

RRosio opened the pull request:
Frontend server call functions and base for backend sync handler and tests

martindurant commented on jupyter_fsspec/handlers.py:
```diff
@@ -251,6 +251,77 @@ def post(self):
+        self.fs_manager = fs_manager
+
+    async def get(self):
+        # remote to local (fetching latest changes)
```
If it's to be explicitly "remote to local", then fs.get() may be simpler, unless you really use features from rsync (e.g., testing the file sizes of files, deleting ones that disappeared on remote).

martindurant commented on jupyter_fsspec/handlers.py:
```diff
@@ -251,6 +251,77 @@ def post(self):
+
+        try:
+            # rsync
+            # (remote_source_path, local_destination_path)
```
So you don't trigger this yet?

martindurant replied:
We can probably figure out here is this is going to be async; OR, we can use the wrapper to make everything async (run on a thread) to make sure we don't block.

martindurant commented on jupyter_fsspec/handlers.py:
```diff
@@ -251,6 +251,77 @@ def post(self):
+# ====================================================================================
+# Handle Syncing Local and Remote filesystems
+# ====================================================================================
+class FilesystemSyncHandler(BaseFileSystemHandler):
```
I would suggest that the FIlesystemManager should have the core functionality, and that the handler only take care of passing on the appropriate parameters and set the output status.

martindurant commented on jupyter_fsspec/handlers.py:
```diff
@@ -251,6 +251,77 @@ def post(self):
+    def initialize(self, fs_manager):
+        self.fs_manager = fs_manager
+
+    async def get(self):
```
`get` implies no change of state. I would call both methods "post", but have them on different endpoints or with different operation parameters as with the other filesystem commands.

martindurant commented on jupyter_fsspec/handlers.py:
```diff
@@ -251,6 +251,77 @@ def post(self):
+    async def get(self):
+        # remote to local (fetching latest changes)
+        request_data = json.loads(self.request.body.decode("utf-8"))
+        local_destination_path = request_data.get("local_path")
```
Were we ever thinking of using pydantic? Unwrapping all the keys and validating seems a little tedious. 
The endpoint doc must document verbosely what the expected input is like (e.g., this might eventually appear in swagger generated docs).

martindurant commented on jupyter_fsspec/handlers.py:
```diff
@@ -556,13 +627,15 @@ def setup_handlers(web_app):
     route_fs_file_transfer = url_path_join(
         base_url, "jupyter_fsspec", "files", "transfer"
     )
+    route_fs_sync = url_path_join(base_url, "jupyter_fsspec", "sync")
```
I don't really see why we need a separate URL here is we don't generally for other operations.

martindurant commented on jupyter_fsspec/tests/test_api.py:
```diff
@@ -539,3 +539,70 @@ async def test_upload_download(fs_manager_instance, jp_fetch):
+    fs_manager = fs_manager_instance
+    remote_fs_info = fs_manager.get_filesystem_by_protocol("s3")
+    remote_key = remote_fs_info["key"]
+    remote_fs = remote_fs_info["info"]["instance"]
```
It's weird to me, that the call to get_filesystem succeeds without an exception but returns nothing.
Also, didn't we key these things by configured name rather than protocol?

martindurant commented on src/handler/fileOperations.ts:
```diff
@@ -366,6 +366,52 @@ export class FsspecModel {
     }
   }
 
+  async sync_push(
```
We may wish to eventually provide a feedback mechanism (see e.g., https://filesystem-spec.readthedocs.io/en/latest/features.html#callbacks ).

martindurant reviewed (commented):
Some thought on this - I wasn't sure if you wanted this already while WIP.

RRosio commented:
Thank you Martin for your feedback! I have worked on some updates:

- Basis for sync with `get` and `put` (as an upload/download from root)
- Removed `BaseFileSystemHandler` => FileSystemManager handlers functionality previously here
- Using single POST handler function for both mechanisms
- Using Pydantic for request data validation
- Using `get_filesystem(key)` ; Removed `get_filesystem_by_protocol()`

Still TBD:
- [ ] Managing the handler endpoint
- [ ] Unclosed client session error on tests => split up upload/download sync and verify/document behavior as expected
	- [ ] Resolve dirty state in between tests with s3 filesystem interaction
- [ ] Provide a Feedback Mechanism

RRosio commented:
I think that this PR is in a good state to merge in and further iterate on. I will go ahead and merge these changes and continue work as a part of the follow-up tickets that have been linked in my comment above!
